File: 0203-Remove Linked List Elements/golang/solution.go
package golang

type ListNode struct {
	Val  int
	Next *ListNode
}

func removeElements(head *ListNode, val int) *ListNode {
	if head == nil {
		return nil
	}
	for head != nil && head.Val == val {
		head = head.Next
	}

	prev := new(ListNode)
	start := head

	for start != nil {
		if start.Val == val {
			prev.Next = start.Next
		} else {
			prev = start
		}
		start = start.Next
	}
	return head
}

// Runtime: 0 ms, faster than 100.00% of Go online submissions for Remove Linked List Elements.
// Memory Usage: 4.7 MB, less than 81.88% of Go online submissions for Remove Linked List Elements.

File: 0203-Remove Linked List Elements/c++/solution.cpp
#include <bits/stdc++.h>
using namespace std;

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeElements(ListNode* head, int val) {
        if(!head) return nullptr;
        while(head != nullptr && head->val == val) head = head->next;
        
        ListNode *prev = nullptr;
        ListNode *start = head;
        while(start) {
            if(start->val == val) {
                // do something to delete node;
                prev->next = start->next;
                start = start->next;
                continue;
            }
            prev = start;
            start = start->next;
        }
        return head;
    }
};

// Runtime: 16 ms, faster than 99.87% of C++ online submissions for Remove Linked List Elements.
// Memory Usage: 13.5 MB, less than 74.39% of C++ online submissions for Remove Linked List Elements.